File: AccuPay/Domain/Paystubs/Allowance.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("employeeallowance")>
    Public Class Allowance

        Public Const FREQUENCY_ONE_TIME As String = "One time"

        Public Const FREQUENCY_DAILY As String = "Daily"

        Public Const FREQUENCY_SEMI_MONTHLY As String = "Semi-monthly"

        Public Const FREQUENCY_MONTHLY As String = "Monthly"

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer?

        Public Property OrganizationID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        Public Property EmployeeID As Integer?

        Public Property ProductID As Integer?

        Public Property EffectiveStartDate As Date

        Public Property AllowanceFrequency As String

        Public Property EffectiveEndDate As Date?

        <NotMapped>
        Public Property TaxableFlag As Char

        <Column("AllowanceAmount")>
        Public Property Amount As Decimal

        <ForeignKey("ProductID")>
        Public Overridable Property Product As Product

        <NotMapped>
        Public Property Type As String
            Get
                Return Product?.PartNo
            End Get
            Set(value As String)
                If Product IsNot Nothing Then Product.PartNo = value
            End Set
        End Property

        Public ReadOnly Property IsOneTime As Boolean
            Get
                Return AllowanceFrequency = FREQUENCY_ONE_TIME
            End Get
        End Property

        Public ReadOnly Property IsDaily As Boolean
            Get
                Return AllowanceFrequency = FREQUENCY_DAILY
            End Get
        End Property

        Public ReadOnly Property IsSemiMonthly As Boolean
            Get
                Return AllowanceFrequency = FREQUENCY_SEMI_MONTHLY
            End Get
        End Property

        Public ReadOnly Property IsMonthly As Boolean
            Get
                Return AllowanceFrequency = FREQUENCY_MONTHLY
            End Get
        End Property

    End Class

End Namespace

File: Accupay.Data/Entities/Allowance.cs
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("employeeallowance")]
    public class Allowance : EmployeeDataEntity
    {
        public const string FREQUENCY_ONE_TIME = "One time";

        public const string FREQUENCY_DAILY = "Daily";

        public const string FREQUENCY_SEMI_MONTHLY = "Semi-monthly";

        public const string FREQUENCY_MONTHLY = "Monthly";

        public int? ProductID { get; set; }

        public DateTime EffectiveStartDate { get; set; }

        public string AllowanceFrequency { get; set; }

        public DateTime? EffectiveEndDate { get; set; }

        [Column("AllowanceAmount")]
        public decimal Amount { get; set; }

        public int? AllowanceTypeId { get; set; }

        [ForeignKey("ProductID")]
        public virtual Product Product { get; set; }

        [ForeignKey("EmployeeID")]
        public virtual Employee Employee { get; set; }

        [ForeignKey("AllowanceTypeId")]
        public virtual AllowanceType AllowanceType { get; set; }

        [NotMapped]
        public string Type
        {
            get => Product?.PartNo != null ? Product?.PartNo : AllowanceType?.Name;
            set
            {
                if (Product != null)
                    Product.PartNo = value;
            }
        }

        public bool IsOneTime => AllowanceFrequency == FREQUENCY_ONE_TIME;

        public bool IsDaily => AllowanceFrequency == FREQUENCY_DAILY;

        public bool IsSemiMonthly => AllowanceFrequency == FREQUENCY_SEMI_MONTHLY;

        public bool IsMonthly => AllowanceFrequency == FREQUENCY_MONTHLY;
    }
}
